torch/nn/utils/parametrize.py

```diff
@@ -53,72 +55,211 @@ def cached():
+            _register_parameter_or_buffer(self, "original", original)
+        else:
+            for i, originali in enumerate(new):
+                if not isinstance(originali, Tensor):
+                    raise ValueError("'right_inverse' must return a Tensor or a Sequence of tensors "
+                                     "(list, tuple...). "
+                                     f"Got element {i} of the sequence with type {type(originali).__name__}.")
+
+                # If the original tensor was a Parameter that required grad, we expect the user to
+                # add the new parameters to the optimizer after registering the parametrization
+                # (this is documented)
+                if isinstance(original, Parameter):
+                    originali = Parameter(originali)
+                originali.requires_grad_(original.requires_grad)
+                _register_parameter_or_buffer(self, f"original{i}", originali)
+
+        # Consistency checks:
+        # Since f : A -> B, right_inverse : B -> A, Z and original should live in B
+        # Z = forward(right_inverse(original))
+        Z = self()
+        if not isinstance(Z, Tensor):
+            raise ValueError(
+                f"A parametrization must return a tensor. Got {type(Z).__name__}."
+            )
+        if Z.dtype != original_dtype:
+            raise ValueError(
+                "Registering a parametrization may not change the dtype of the tensor.\n"
+                f"unparametrized dtype: {original_dtype}\n"
+                f"parametrized dtype: {Z.dtype}"
+            )
+        if Z.shape != original_shape:
+            raise ValueError(
+                "Registering a parametrization may not change the shape of the tensor.\n"
+                f"unarametrized shape: {original_shape}\n"
+                f"parametrized shape: {Z.shape}"
+            )
 
-        It calls the methods ``right_inverse`` (see :func:`register_parametrization`)
-        of the parametrizations in the inverse order that they have been registered.
-        Then, it assigns the result to ``self.original``.
+    def right_inverse(self, value: Tensor) -> None:
+        r"""Calls the methods ``right_inverse`` (see :func:`register_parametrization`)
+        of the parametrizations in the inverse order they were registered in.
+        Then, it stores the result in ``self.original`` if ``right_inverse`` outputs one tensor
+        or in ``self.original0``, ``self.original1``, ... if it outputs several.
 
         Args:
             value (Tensor): Value to which initialize the module
-
-        Raises:
-            RuntimeError: if any of the parametrizations do not implement a ``right_inverse`` method
         """
+        # All the exceptions in this function should almost never throw.
+        # They could throw if, for example, right_inverse function does not return the same dtype
+        # for every input, which should most likely be caused by a bug in the code
```

Comment:
nit: maybe this could be clearer if we say that "right_inverse function returns a different dtype when given a different input"